src/resdk/resources/metadata.py

```diff
@@ -0,0 +1,232 @@
+        return self._df_bytes
+
+    def validate_df(self, df):
+        """Validate fd property.
+
+        Validates that df:
+
+        - is an instance of pandas.DataFrame
+        - has one of the columns ID / slug / name
+        - Values in that column matches at least some samples
+
+            - If not matches, raise warning
+            - If there are samples in df but not in collection, raise warning
+            - If there are samples in collection but not in df, raise warning
+
+        """
+        if not isinstance(df, pd.DataFrame):
+            raise ValueError("Attribute ad must be a pandas.DataFrame object.")
+
+        for match_column in self.sample_identifier_columns:
+            if match_column in df.columns:
+                break
+        else:
+            options = ", ".join(self.sample_identifier_columns)
+            raise ValueError(
+                f"There should be a column in df with one of the following names: {options}"
+            )
+
+        # Sample identifiers from df
+        # Convert to string to avoid string vs. integer comparisons
+        df_samples = set(df[match_column].astype(str).tolist())
+
+        # Sample identifiers from collection
+        # Convert to string to avoid string vs. integer comparisons
+        attr = self.sample_identifier_columns[match_column]
+        col_samples = {str(getattr(s, attr)) for s in self.collection.samples}
+
+        intersection = df_samples & col_samples
+        if not intersection:
+            warnings.warn(
+                "No intersection between samples in df and samples in collection."
+            )
+
+        not_in_col = df_samples - col_samples
+        if not_in_col:
+            not_in_col = ", ".join(list(not_in_col))
+            warnings.warn(
+                f"There are samples in df that are not in collection: {not_in_col}"
+            )
+
+        not_in_df = col_samples - df_samples
+        if not_in_df:
+            not_in_df = ", ".join(list(not_in_df))
+            warnings.warn(
+                f"There are samples in collection that are not in df: {not_in_df}"
+            )
+
+    def get_df(self, parser=None, **kwargs):
+        """Get table as pd.DataFrame."""
+        # Do not use cached value if parser is specified
+        if self._df is None or parser is not None:
+
+            # Enable parsing the byte stream with arbitrary parser, not just pandas
+            # Otherwise try to guess the parser based on file extension
+            basename = self.output["table"]["file"]
+            if parser is not None:
+                df = parser(self.df_bytes, **kwargs)
+            elif basename.endswith("xls"):
+                df = pd.read_excel(self.df_bytes, engine="xlrd")
+            elif basename.endswith("xlsx"):
+                df = pd.read_excel(self.df_bytes, engine="openpyxl")
+            elif any(basename.endswith(ext) for ext in ["tab", "tsv"]):
+                df = pd.read_csv(self.df_bytes, sep="\t")

```

Comment:
We should be mindful of float precision and rounding when loading data; [see this example](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#specifying-method-for-floating-point-conversion). Say we get the data frame from collection, do some modifications and save it afterward. This could introduce undesired changes to the metadata. I am not sure how often this can happen but users should at least be made aware of it.

Reply:
Thanks for pointing this out, added "round_trip" option for float precision.